File: selection_sort.py
# Time: O(n^2) | Space: O(1)
def selectionSort(array):
    for currentIdx in range(len(array)):
        smallestIdx = currentIdx
        for j in range(currentIdx+1, len(array)):
            if array[smallestIdx] > array[j]:
                smallestIdx = j
        swap(currentIdx, smallestIdx, array)

    return array

def swap(i, j, array):
    array[i], array[j] = array[j], array[i] 

if __name__ == "__main__":
    array = [8, 2, 5, 4, 1]
    print(selectionSort(array))

# output: [1, 2, 4, 5, 8]

File: selection_sort.go
package main

import "fmt"

// Time: O(n^2) | Space: O(1)
func selectionSort(array []int) []int {
	for currentIdx := 0; currentIdx < len(array); currentIdx++ {
		smallestIdx := currentIdx
		for j := currentIdx + 1; j < len(array); j++ {
			if array[smallestIdx] > array[j] {
				smallestIdx = j
			}
		}
		swap(currentIdx, smallestIdx, array)
	}
	return array
}

func swap(i, j int, array []int) {
	array[i], array[j] = array[j], array[i]
}

func main() {
	array := []int{8, 2, 5, 4, 1}
	fmt.Println(selectionSort(array))
}

/* output:
[1 2 4 5 8]
*/
